azure-keyvault/azure/keyvault/custom/internal.py

```diff
@@ -0,0 +1,225 @@
+#---------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+#---------------------------------------------------------------------------------------------
+
+import hmac
+import hashlib
+import struct
+import json
+import codecs
+from base64 import b64encode, b64decode
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+from cryptography.hazmat.backends import default_backend
+
+
+def _a128cbc_hs256_encrypt(key, iv, plaintext, authdata):
+    if not key or not len(key) >= 32:
+        raise ValueError('key must be at least 256 bits for algorithm "A128CBC-HS256"')
+    if not iv or len(iv) != 16:
+        raise ValueError('iv must be 128 bits for algorithm "A128CBC-HS256"')
+    if not plaintext:
+        raise ValueError('plaintext must be specified')
+    if not authdata:
+        raise ValueError('authdata must be specified')
+
+    # get the hmac key and the aes key from the specified key
+    hmac_key = key[:16]
+    aes_key = key[16:32]
+
+    # calculate the length of authdata and store as bytes
+    auth_data_length = _int_to_bigendian_8_bytes(len(authdata) * 8)
+
+    # pad the data so it is a multiple of blocksize
+    # pkcs7 padding is expected to the pad bytes are the value of the number of pad bytes added
+    padlen = 16 - len(plaintext) % 16
+    padval = _int_to_bytes(padlen)
+    plaintext += padval * padlen
+
+    # create the cipher and encrypt the plaintext
+    cipher = Cipher(algorithms.AES(aes_key), modes.CBC(iv), backend=default_backend())
+    encryptor = cipher.encryptor()
+    ciphertext = encryptor.update(plaintext) + encryptor.finalize()
+
+    # get the data to hash with HMAC, hash the data and take the first 16 bytes
+    hashdata = authdata + iv + ciphertext + auth_data_length
+    tag = hmac.new(key=hmac_key, msg=hashdata, digestmod=hashlib.sha256).digest()[:16]
+
+    return ciphertext, tag
+
+
+def _a128cbc_hs256_decrypt(key, iv, ciphertext, authdata, authtag):
+    if not key or not len(key) >= 32:
+        raise ValueError('key must be at least 256 bits for algorithm "A128CBC-HS256"')
+    if not iv or len(iv) != 16:
+        raise ValueError('iv must be 128 bits for algorithm "A128CBC-HS256"')
+    if not ciphertext:
+        raise ValueError('ciphertext must be specified')
+    if not authdata:
+        raise ValueError('authdata must be specified')
+
+    hmac_key = key[:16]
+    aes_key = key[16:32]
+    auth_data_length = _int_to_bigendian_8_bytes(len(authdata) * 8)
+
+    hashdata = authdata + iv + ciphertext + auth_data_length
+    tag = hmac.new(key=hmac_key, msg=hashdata, digestmod=hashlib.sha256).digest()[:16]
+
+    if tag != authtag:
```

Comment:
Consider constant time compare for this operation, i.e. force maximum compare. The authtag length could also be checked earlier to ensure that it matches the expected size for SHA256.